tests/cpp/test_memory.cpp

```diff
@@ -668,6 +679,89 @@ TEST_F(TMAIndexingTest, NonTrivialGmemAllocationDomain) {
 
 class TMAMiscTest : public TMATest {};
 
+TEST_F(TMAMiscTest, AdvancedThreadParallelizationLoad) {
```

Comment:
Kernel:

```CUDA
__global__ void nvfuser_none_f0_c0_r0_g0(Tensor<float, 1, 1> T0, const __grid_constant__ TensorMap var0, Tensor<float, 1, 1> T2) {
  alignas(16) extern __shared__ char array[];
  const unsigned smem_offset = 0;
  const TensorMap* ptr1;
  ptr1 = &var0;
  nvfuser_index_t i2;
  i2 = 512LL * ((nvfuser_index_t)blockIdx.x);
  Array<nvfuser_index_t, 1, 1> a3;
  a3 = Array<nvfuser_index_t, 1, 1>{((128LL * ((nvfuser_index_t)threadIdx.x)) + i2)};
  nvfuser_index_t i4;
  i4 = ((nvfuser_index_t)threadIdx.x) + i2;
  float* T1 = reinterpret_cast<float*>(array + smem_offset + 0);
  uint64_t* T3 = reinterpret_cast<uint64_t*>(array + smem_offset + 2048LL);
  mbarrier::init(toSmem(T3), 1U);
  __syncthreads();
  if ((((nvfuser_index_t)threadIdx.x) < 4)) {
    uint64_t i5;
    i5 = mbarrier::arriveExpectTX(toSmem(T3), 512U);
    Hopper::cpAsyncBulkTensorTileG2S((Hopper::CpAsyncBulkTensorTileG2SIndex<1>{ ptr1, a3, toSmem(T3) }), (toSmem(T1) + (512 * ((nvfuser_index_t)threadIdx.x))));
    mbarrier::wait(toSmem(T3), i5);
  }
  __syncthreads();
  mbarrier::inval(toSmem(T3));
  __syncthreads();
  if ((i4 < T0.logical_size[0LL])) {
    T2[i4]
       = T1[((nvfuser_index_t)threadIdx.x)];
  }
}
```

Note that this kernel will not work.

Reply:
I'm not sure if it is performance wise good, that mbarrier object in smem is effectively initialized by all thread in a block:
```
mbarrier::init(toSmem(T3), 4U);
```
In the next step 4 threads will set target for expected size of memory transaction (in smem) in the same token.

After that we use the same token object (T3, in smem) to control 4 different memory transactions with parts of T1 tensor.
```
Hopper::cpAsyncBulkTensorTileG2S((Hopper::CpAsyncBulkTensorTileG2SIndex<1>{ ptr1, a3, toSmem(T3) }), (toSmem(T1) + (512 * ((nvfuser_index_t)threadIdx.x))));
```
Won't we have some UB / race with such implementation?

Reply:
Good question! And my answer is: I don't know. I don't know if it is defined behavior to initialize the mbarrier multiple times in multiple threads, I also don't know if it is defined behavior for each thread to wait for its own token, instead of having a block-wide token. In practice, this code works. So I will leave it as is, because the current synchronization is just a hack to make things works, we will have to change it in the future anyway.

Reply:
Ok, got it - I'm really surprised this worked, to be honest.

So let's keep it as it is.